components/layout/table_row.rs

```diff
@@ -107,96 +106,158 @@ impl TableRowFlow {
+                if *span <= 1 {
+                    break;
                 }
-
-                {
-                    let child_fragment = kid.as_mut_table_cell().fragment();
-                    // TODO: Percentage block-size
-                    let child_specified_block_size =
-                        MaybeAuto::from_style(child_fragment.style().content_block_size(),
-                                              Au(0)).specified_or_zero();
-                    max_block_size =
-                        max(max_block_size,
-                            child_specified_block_size +
-                            child_fragment.border_padding.block_start_end());
+                let incoming = incoming_rowspan_data[*col];
+                *max_block_size = max(*max_block_size, incoming);
+                if *span > 2 {
+                    *largest_leftover_incoming_size = max(*largest_leftover_incoming_size,
+                                                          incoming * (*span - 1) as i32)
                 }
-                let child_node = kid.mut_base();
-                child_node.position.start.b = Au(0);
-                max_block_size = max(max_block_size, child_node.position.size.block);
+                *col += 1;
             }
+        }
+        // Per CSS 2.1 § 17.5.3, find max_y = max(computed `block-size`, minimum block-size of
+        // all cells).
+        let mut max_block_size = Au(0);
+        let mut largest_leftover_incoming_size = Au(0);
+        let thread_id = self.block_flow.base.thread_id;
+        let content_box = self.block_flow.base.position
+            - self.block_flow.fragment.border_padding
+            - self.block_flow.fragment.margin;
 
-            let mut block_size = max_block_size;
-            // TODO: Percentage block-size
-            block_size = match MaybeAuto::from_style(self.block_flow
-                                                         .fragment
-                                                         .style()
-                                                         .content_block_size(),
-                                                     Au(0)) {
-                MaybeAuto::Auto => block_size,
-                MaybeAuto::Specified(value) => max(value, block_size),
-            };
+        let mut col = 0;
+        for kid in self.block_flow.base.child_iter_mut() {
+            include_sizes_from_previous_rows(&mut col, &self.incoming_rowspan,
+                                             incoming_rowspan_data, &mut max_block_size,
+                                             &mut largest_leftover_incoming_size);
+            kid.place_float_if_applicable();
+            if !kid.base().flags.is_float() {
+                kid.assign_block_size_for_inorder_child_if_necessary(layout_context,
+                                                                     thread_id,
+                                                                     content_box);
+            }
```

Comment:
While you're here, can you change this `kid.base().flags.is_float()` condition to a debug assert? Children of table rows should never be floating.